File: python/webhooks/webhook_receiver.py
#!/usr/bin/env python3
"""
Webhook Receiver with HMAC-SHA256 Verification - Veritier Example (Python)
===========================================================================
Verifies X-Veritier-Signature against the raw body bytes, then processes
extract / verify / validate / attestation deliveries.

Retries send the same Idempotency-Key / X-Veritier-Idempotency-Key
(attest: attestation:{action_id}:{transaction_id}). HMAC is over the body only.

Setup:
  1. pip install flask python-dotenv
  2. Set VERITIER_WEBHOOK_SECRET in your .env (vtsec_... from the dashboard)
  3. python webhook_receiver.py
  4. Listen on http://localhost:5050/webhooks/veritier

Dashboard allows http://localhost for test webhooks. Production URLs need HTTPS.

Full webhook docs: https://veritier.ai/docs#webhooks
"""

import hmac
import hashlib
import os
from dotenv import load_dotenv
from flask import Flask, request, abort, jsonify

load_dotenv()

app = Flask(__name__)

WEBHOOK_SECRET = os.getenv("VERITIER_WEBHOOK_SECRET", "")
_seen_idem = set()

if not WEBHOOK_SECRET:
    print("Warning: VERITIER_WEBHOOK_SECRET is not set.")
    print("  Configure a webhook at https://veritier.ai/dashboard to get your secret.")


@app.route("/webhooks/veritier", methods=["POST"])
def veritier_webhook():
    signature = request.headers.get("X-Veritier-Signature", "")
    idem = request.headers.get("Idempotency-Key") or request.headers.get(
        "X-Veritier-Idempotency-Key", ""
    )
    action_id = request.headers.get("X-Veritier-Action-Id", "")

    if not WEBHOOK_SECRET:
        print("Webhook secret not configured - rejecting request")
        abort(500)

    raw_body = request.data
    expected_signature = "vtsec_" + hmac.new(
        key=WEBHOOK_SECRET.encode("utf-8"),
        msg=raw_body,
        digestmod=hashlib.sha256,
    ).hexdigest()

    if not hmac.compare_digest(signature, expected_signature):
        print("Invalid webhook signature - rejecting request")
        abort(401)

    if idem:
        if idem in _seen_idem:
            print(f"Duplicate Idempotency-Key {idem} - ignoring retry")
            return jsonify({"status": "ok", "duplicate": True}), 200
        _seen_idem.add(idem)

    payload = request.get_json(force=True)
    transaction_id = payload.get("transaction_id", "unknown")
    kind = payload.get("type", "event")
    is_test = payload.get("is_test", False)
    results = payload.get("results")

    print(f"\n{'─' * 50}")
    if is_test:
        print("[TEST MODE] No quota was consumed.")
    print(f"type={kind}  tx={transaction_id}  idem={idem or '(none)'}")
    if action_id:
        print(f"X-Veritier-Action-Id={action_id}")

    if kind == "attestation" and isinstance(results, dict):
        print(f"  decision={results.get('decision')}  credential_id={results.get('credential_id')}")
        print(f"  action_id={payload.get('action_id') or results.get('action_id')}")
    elif isinstance(results, list):
        print(f"  claims: {len(results)}")
        for res in results:
            print(f"    {res.get('claim')} -> {res.get('verdict')}")
    else:
        print(f"  results: {str(results)[:200]}")

    print(f"{'─' * 50}\n")
    return jsonify({"status": "ok"}), 200


@app.route("/health", methods=["GET"])
def health():
    return jsonify({"status": "ok"}), 200


if __name__ == "__main__":
    print("Veritier Webhook Receiver")
    print("  Listening on http://localhost:5050/webhooks/veritier")
    app.run(host="0.0.0.0", port=5050, debug=True)

File: javascript/webhooks/webhook_receiver.mjs
/**
 * Webhook Receiver with HMAC-SHA256 Verification - Veritier Example (JavaScript)
 * ================================================================================
 * Verifies X-Veritier-Signature against the raw body bytes, then processes
 * extract / verify / validate / attestation deliveries.
 *
 * Retries send the same Idempotency-Key / X-Veritier-Idempotency-Key
 * (attest: attestation:{action_id}:{transaction_id}). HMAC is over the body only.
 *
 * Setup:
 *   1. npm install
 *   2. Set VERITIER_WEBHOOK_SECRET in .env
 *   3. node webhook_receiver.mjs
 *
 * Docs: https://veritier.ai/docs#webhooks
 */

import "dotenv/config";
import express from "express";
import { createHmac, timingSafeEqual } from "node:crypto";

const app = express();
const PORT = 5050;
const WEBHOOK_SECRET = process.env.VERITIER_WEBHOOK_SECRET || "";
const seenIdem = new Set();

if (!WEBHOOK_SECRET) {
  console.warn("Warning: VERITIER_WEBHOOK_SECRET is not set.");
}

app.use("/webhooks/veritier", express.raw({ type: "application/json" }));
app.use(express.json());

app.post("/webhooks/veritier", (req, res) => {
  const signature = req.headers["x-veritier-signature"] || "";
  const idem =
    req.headers["idempotency-key"] || req.headers["x-veritier-idempotency-key"] || "";
  const actionId = req.headers["x-veritier-action-id"] || "";

  if (!WEBHOOK_SECRET) {
    return res.status(500).json({ error: "Webhook secret not configured" });
  }

  const rawBody = req.body;
  const expectedSignature =
    "vtsec_" + createHmac("sha256", WEBHOOK_SECRET).update(rawBody).digest("hex");

  try {
    const sigBuffer = Buffer.from(signature, "utf8");
    const expectedBuffer = Buffer.from(expectedSignature, "utf8");
    if (
      sigBuffer.length !== expectedBuffer.length ||
      !timingSafeEqual(sigBuffer, expectedBuffer)
    ) {
      return res.status(401).json({ error: "Invalid signature" });
    }
  } catch {
    return res.status(401).json({ error: "Invalid signature" });
  }

  if (idem) {
    if (seenIdem.has(idem)) {
      console.log(`Duplicate Idempotency-Key ${idem} - ignoring retry`);
      return res.json({ status: "ok", duplicate: true });
    }
    seenIdem.add(idem);
  }

  const payload = JSON.parse(rawBody.toString("utf8"));
  const transactionId = payload.transaction_id || "unknown";
  const kind = payload.type || "event";
  const results = payload.results;

  console.log(`\n${"─".repeat(50)}`);
  if (payload.is_test) console.log("[TEST MODE] No quota was consumed.");
  console.log(`type=${kind}  tx=${transactionId}  idem=${idem || "(none)"}`);
  if (actionId) console.log(`X-Veritier-Action-Id=${actionId}`);

  if (kind === "attestation" && results && !Array.isArray(results)) {
    console.log(`  decision=${results.decision}  credential_id=${results.credential_id}`);
  } else if (Array.isArray(results)) {
    console.log(`  claims: ${results.length}`);
    for (const r of results) {
      console.log(`    ${r.claim} -> ${r.verdict}`);
    }
  }

  console.log(`${"─".repeat(50)}\n`);
  return res.json({ status: "ok" });
});

app.get("/health", (_req, res) => {
  res.json({ status: "ok" });
});

app.listen(PORT, () => {
  console.log("Veritier Webhook Receiver");
  console.log(`  Listening on http://localhost:${PORT}/webhooks/veritier`);
});
